### ISSUE TITLE
Notice when two apps overwrite the same static file path

### ISSUE THREAD

**Issue by warpech** [Starcounter/Starcounter#3962]
Improvement suggestion: 

When I start app A, which has a static file, for example:

```
/sys/paper-input/paper-input.html
```

And then I start app B, which registers a static file at the same path:

```
/sys/paper-input/paper-input.html
```

The app B will take priority over app A. The problem is that I don't notice that and the developer might spend hours debugging a wrong thing.

I suggest we add a notice in Starcounter log that app B overwrites a static file from app A, with the path to that file.

Caution: This may result in many notice, because sometimes a Web Component has 50+ static files (language packs, docs, etc).

@miyconst, @Mnosek, @aloscha, @alemoi WDYT?

**Comment by miyconst** [comment]
>I suggest we add a notice in Starcounter log that app B overwrites a static file from app A, with the path to that file.

I don't think that will help much, since developers very often do not look into the log. I would update Starcounter Administrator `Public web folders` tab to provide more info.

**Comment by ubbeK** [comment]
Why not pipe a yellow colored warning to the console when running `star.exe B.exe` in addition to @miyconst suggestion.

**Comment by warpech** [comment]
> I don't think that will help much, since developers very often do not look into the log.

Maybe they will get into the habit when useful stuff like this appears in the log.

> Why not pipe a yellow colored warning to the console when running star.exe B.exe in addition to @miyconst suggestion.

This would be nice, too. But it won't work for apps started from Administrator/Warehouse.

**Comment by warpech** [comment]
I think that all 3 ideas presented in this issue are good. They are not excluding.

**Comment by aloscha** [comment]
It helps of course, but this solution doesn't solve the problem. What if we have access dynamically?
For example `paper-input` in A is 1.0.0 and in B is 1.0.1, than we crate mirror in `/sys/paper-input-1.0.0` and `/sys/paper-input-1.0.1` and dynamically apps will get proper lib?

Because really now it is generating a lot of problems (we have access to all sourcecode), but what happens if I would like to install some GraphControl with `paper-input 0.9` - I need to manage priority of that.

**Comment by per-samuelsson** [comment]
> This would be nice, too. But it won't work for apps started from Administrator/Warehouse.

Why not? The console is redirected and show in the web Administrator.

**Comment by warpech** [comment]
> It helps of course, but this solution doesn't solve the problem. What if we have access dynamically?
For example `paper-input` in A is 1.0.0 and in B is 1.0.1, than we crate mirror in `/sys/paper-input-1.0.0` and `/sys/paper-input-1.0.1` and dynamically apps will get proper lib?
> Because really now it is generating a lot of problems (we have access to all sourcecode), but what happens if I would like to install some GraphControl with `paper-input 0.9` - I need to manage priority of that.

You would also need to rename `<paper-input>` in HTML, JS and CSS code to `<paper-input-1.0.1>` to make this work. Repeat this for all dependencies. 

I am afraid there is no chance that automatic code rewriting could do this correctly.

**Comment by aloscha** [comment]
>I am afraid there is no chance that automatic code rewriting could do this correctly.

What if put starcounter-include into <iframe/> like https://github.com/Juicy/juicy-iframe? Each webcomponents will be available just in current iframe and not for whole project?

cc @tomalec

**Comment by warpech** [comment]
> What if put starcounter-include into <iframe/> like https://github.com/Juicy/juicy-iframe? Each webcomponents will be available just in current iframe and not for whole project? So we don't have to change version or path.

The whole point of Web Components is not having to use iframe. This is like asking to put a gasoline engine in a Tesla car to extend its range ;)

**Comment by SenerDemiral** [comment]
```
This is like asking to put a gasoline engine in a Tesla car to extend its range.
```
Awesome analogy.

**Comment by aloscha** [comment]
I get it, but putting iframe solves the problem - for now, maybe later we will create some sandboxing.

You need to travel from LosAngeles to Phoenix using your Tesla (600km) and this is matter of life or death, so you will put gasoline engine into Tesla to extend range, otherwise you don't get into finish line.

Sum up if adding iframe will fix the problem and we save a lot of time why don't to use it?

**Comment by warpech** [comment]
Mostly 2 reasons:

1. Using iframe to load partial from another app would make it non-blendable by starcounter-composition.
2. Iframes are slow

If that is a matter of life and death, then you can use an iframe. But this thread is not about such case.

**Comment by jwillmer** [comment]
I get your reluctance of iframes but I a simple warning does not help if you like to use apps from different vendors. 

> I don't like to change/update my stable version of my app A only because you like to use the unstable(?) new version of a lib that I also use.

Does someone have a better idea then iframes? There needs to be something better?!
